### PR TITLE
fix(source-jira): upgrade to CDK v6 to use concurrency for incremental streams

### PR DESCRIPTION
**Pull request by tolik0**
## What
This PR upgrades the CDK to version 6 and leverages its concurrency capabilities to enhance the handling of incremental streams. Additionally, it introduces substream state migration to ensure reliable data synchronization.

## How
**CDK Upgrade & Concurrency Enhancements:**

- Upgraded CDK from version 5 to version 6 to support concurrent processing of multiple incremental streams, enhancing performance and efficiency.
- Updated configuration files and dependencies to accommodate changes introduced in CDK v6.

**Substream State Migration:**

- Implemented state migration logic within the `PerPartitionWithFallback` class to handle incremental dependencies.
- Modified the `set_initial_state` method to ensure that parent streams correctly inherit state from child streams during the initial sync.

## User Impact
Fix error with substreams failing with retry error after the previous update because of rereading parent stream.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml [hunk 1]
```diff
@@ -1416,3 +1416,8 @@ check:
     - board_issues
     - issues
     - sprint_issues
+
+concurrency_level:
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:1420]: Can we add a comment saying that this section is useless for now as Jira as no concurrent streams now the reason being that only datetimebasedcursor without partition_routers are supported

File: airbyte-integrations/connectors/source-jira/source_jira/run.py [hunk 2]
```diff
@@ -1,14 +1,53 @@
...
+                            error=AirbyteErrorTraceMessage(
+                                message=f"Error starting the sync. This could be due to an invalid configuration or catalog. Please contact Support for assistance. Error: {error}",
+                                stack_trace=traceback.format_exc(),
+                            ),
+                        ),
+                    )
+                )
+            ).decode()
+        )
+        return None
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/run.py:45]: Should we re-raise the error here instead? It seems like returning none will just fail later on and therefore it is better to fail clearly here. I updated the comment [here](https://github.com/airbytehq/airbyte-internal-issues/issues/10550#issuecomment-2471128821) to reflect that

File: airbyte-integrations/connectors/source-jira/source_jira/source.py [hunk 3]
```diff
@@ -70,12 +71,12 @@ def _validate_and_transform_config(self, config: Mapping[str, Any]):
 
     @staticmethod
     def get_authenticator(config: Mapping[str, Any]):
-        return BasicHttpAuthenticator(config["email"], config["api_token"])
+        return BasicHttpAuthenticator(config.get("email"), config["api_token"])
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/source.py:74]: Why this change? email is required in the spec. Same comment with `domain` on line 79

**Review by maxi297 — APPROVED** [review]
A couple of comments but I'm preemptively approving

File: airbyte-integrations/connectors/source-jira/source_jira/source.py (hunk 3 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-jira/source_jira/source.py:74]: Yes, it should, but this line is executed before the check, so we fail with:
```
{"type":"TRACE","trace":{"type":"ERROR","emitted_at":1731587941523,"error":{"message":"Error starting the sync. This could be due to an invalid configuration or catalog. Please contact Support for assistance. Error: 'domain'","stack_trace":"Traceback (most recent call last):\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/source_jira/run.py\", line 22, in _get_source\n    return SourceJira(\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/source_jira/source.py\", line 24, in __init__\n    super().__init__(catalog=catalog, config=config, state=state, **{\"path_to_yaml\": \"manifest.yaml\"})\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/.source-jira/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/yaml_declarative_source.py\", line 33, in __init__\n    super().__init__(\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/.source-jira/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/concurrent_declarative_source.py\", line 81, in __init__\n    self._concurrent_streams, self._synchronous_streams = self._group_streams(\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/.source-jira/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/concurrent_declarative_source.py\", line 181, in _group_streams\n    for declarative_stream in self.streams(config=config):\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/source_jira/source.py\", line 62, in streams\n    return streams + self.get_non_portable_streams(config=config)\n  File \"/home/anatolii/airbytehq/airbyte/airbyte-integrations/connectors/source-jira/source_jira/source.py\", line 79, in get_non_portable_streams\n    args = {\"authenticator\": authenticator, \"domain\": config[\"domain\"], \"projects\": config[\"projects\"]}\nKeyError: 'domain'\n"}}}
```
With the get the error is correct:
```
{"type":"CONNECTION_STATUS","connectionStatus":{"status":"FAILED","message":"Config validation error: 'domain' is a required property"}}
```

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml (hunk 1 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:1421]: Fixed

File: airbyte-integrations/connectors/source-jira/source_jira/source.py (hunk 3 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-jira/source_jira/source.py:74]: Ok! So the reason that it was failing properly before is that `AirbyteEntrypoint.validate_connection` was performed after the source is instantiated. Now, the instantiation of the source is failing. This is kind of annoying because to call `validate_connection`, we need the spec and to have the spec we need to instantiate the source. `Source.spec` method should probably be a classmethod but right now it isn't. So we are at risk of having an invalid config being provided an failing with a trace message instead of a connection_status failed.
